### PR TITLE
rpc: avoid leaking threads during checktx

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: internal/rpc/core/mempool.go [hunk 1]
```diff
@@ -64,61 +73,70 @@ func (env *Environment) BroadcastTxCommit(ctx context.Context, tx types.Tx) (*co
 	err := env.Mempool.CheckTx(
 		ctx,
 		tx,
-		func(res *abci.ResponseCheckTx) { resCh <- res },
+		func(res *abci.ResponseCheckTx) {
```
Comment (williambanfield) [inline internal/rpc/core/mempool.go:76]: What would you think of having this callback take a context directly?

File: internal/rpc/core/mempool.go [hunk 2]
```diff
@@ -64,61 +73,70 @@ func (env *Environment) BroadcastTxCommit(ctx context.Context, tx types.Tx) (*co
 	err := env.Mempool.CheckTx(
```
Comment (williambanfield) [inline internal/rpc/core/mempool.go:73]: The way we're writing this function looks a whole lot like a regular function call with extra steps. We want to block on the result of its computation, I'm not totally sure I understand why we don't just write this as a function call that returns the ResponseCheckTx :thinking:

Comment (tychoish) [reply internal/rpc/core/mempool.go:73]: I think that's a good move eventually, but the tests *really* depend on injecting behavior into the call back, so it's harder than it should be to get out of this pickle.

Comment (creachadair) [reply internal/rpc/core/mempool.go:73]: Might be worth adding a TODO to document that this is why we can't easily rip it out (I think we've bumped into this before).

**Review by creachadair — APPROVED** [review]

**Pull request merged**